src/main/java/de/tum/in/www1/artemis/service/connectors/BambooService.java

```diff
@@ -774,6 +775,35 @@ private Long calculateScoreForResult(Result result) {
         return (long) 0;
     }
 
+    @Override
+    @Transactional
+    public Optional<Result> retrieveLatestBuildResult(ProgrammingExerciseParticipation participation, ProgrammingSubmission submission) {
+        Map<String, Object> buildResults = queryLatestBuildResultFromBambooServer(participation.getBuildPlanId());
+        if (buildResults == null) {
+            return Optional.empty();
+        }
+        // The retrieved build result must match the commitHash of the provided submission.
+        String commitHash = (String) buildResults.get("vcsRevisionKey");
+        if(!commitHash.equalsIgnoreCase(submission.getCommitHash())) {
+            return Optional.empty();
+        }
+        Result result = new Result();
+        result.setRatedIfNotExceeded(participation.getProgrammingExercise().getDueDate(), ZonedDateTime.now());
```

Comment:
```suggestion
        result.setRatedIfNotExceeded(participation.getProgrammingExercise().getDueDate(), submission);
```

Reply:
This is a small logic error. We need to use the submission date here, because now might be too late to be considered as rated